File: java/src/KdTree.java
import java.awt.geom.Point2D;
import java.util.*;

public class KdTree {

	public static class Node {
		Point2D.Double p;
		Node left;
		Node right;
	}

	static final Comparator<Point2D.Double> cmpX = new Comparator<Point2D.Double>() {
		public int compare(Point2D.Double p1, Point2D.Double p2) {
			return Double.compare(p1.x, p2.x);
		}
	};

	static final Comparator<Point2D.Double> cmpY = new Comparator<Point2D.Double>() {
		public int compare(Point2D.Double p1, Point2D.Double p2) {
			return Double.compare(p1.y, p2.y);
		}
	};

	Node root;

	public KdTree(Point2D.Double[] points) {
		root = buildTree(0, points.length, true, points);
	}

	Node buildTree(int left, int right, boolean divX, Point2D.Double[] points) {
		if (left >= right)
			return null;
		Arrays.sort(points, left, right, divX ? cmpX : cmpY);
		int mid = (left + right) >> 1;
		Node node = new Node();
		node.p = points[mid];
		node.left = buildTree(left, mid, !divX, points);
		node.right = buildTree(mid + 1, right, !divX, points);
		return node;
	}

	static double EPS = 1e-9;
	double closestDist;
	Node closestNode;

	public Node findNearestNeighbour(double x, double y) {
		closestDist = Double.POSITIVE_INFINITY;
		findNearestNeighbour(root, x, y, true);
		return closestNode;
	}

	void findNearestNeighbour(Node node, double x, double y, boolean divX) {
		if (node == null)
			return;
		double d = Point2D.distanceSq(node.p.x, node.p.y, x, y);
		if (closestDist > d) {
			closestDist = d;
			closestNode = node;
		}
		double delta = divX ? x - node.p.x : y - node.p.y;
		double delta2 = delta * delta;

		Node node1 = delta < 0 ? node.left : node.right;
		Node node2 = delta < 0 ? node.right : node.left;

		findNearestNeighbour(node1, x, y, !divX);
		if (delta2 < closestDist) {
			findNearestNeighbour(node2, x, y, !divX);
		}
	}

	// Usage example
	public static void main(String[] args) {
		double[] x = { 0, 10, 0, 10 };
		double[] y = { 0, 10, 10, 0 };
		Point2D.Double[] points = new Point2D.Double[x.length];
		for (int i = 0; i < points.length; i++) {
			points[i] = new Point2D.Double(x[i], y[i]);
		}
		KdTree kdTree = new KdTree(points);
		KdTree.Node res = kdTree.findNearestNeighbour(6, 3);
		System.out.println(res.p == points[3]);
	}
}

File: cpp/KdTree.cpp
#include <algorithm>
#include <climits>
#include <vector>
#include <iostream>
using namespace std;

typedef pair<int, int> pii;
typedef vector<pii> vpii;

const int maxn = 100000;
int tx[maxn];
int ty[maxn];
bool divX[maxn];

bool cmpX(const pii &a, const pii &b) {
	return a.first < b.first;
}

bool cmpY(const pii &a, const pii &b) {
	return a.second < b.second;
}

void buildTree(int left, int right, pii points[]) {
	if (left >= right)
		return;
	int mid = (left + right) >> 1;

	//sort(points + left, points + right + 1, divX ? cmpX : cmpY);
	int minx = INT_MAX;
	int maxx = INT_MIN;
	int miny = INT_MAX;
	int maxy = INT_MIN;
	for (int i = left; i < right; i++) {
		checkmin(minx, points[i].first);
		checkmax(maxx, points[i].first);
		checkmin(miny, points[i].second);
		checkmax(maxy, points[i].second);
	}
	divX[mid] = (maxx - minx) >= (maxy - miny);
	nth_element(points + left, points + mid, points + right, divX[mid] ? cmpX : cmpY);

	tx[mid] = points[mid].first;
	ty[mid] = points[mid].second;

	if (left + 1 == right)
		return;
	buildTree(left, mid, points);
	buildTree(mid + 1, right, points);
}

long long closestDist;
int closestNode;

void findNearestNeighbour(int left, int right, int x, int y) {
	if (left >= right)
		return;
	int mid = (left + right) >> 1;
	int dx = x - tx[mid];
	int dy = y - ty[mid];
	long long d = dx * (long long) dx + dy * (long long) dy;
	if (closestDist > d && d) {
		closestDist = d;
		closestNode = mid;
	}
	if (left + 1 == right)
		return;

	int delta = divX[mid] ? dx : dy;
	long long delta2 = delta * (long long) delta;
	int l1 = left;
	int r1 = mid;
	int l2 = mid + 1;
	int r2 = right;
	if (delta > 0)
		swap(l1, l2), swap(r1, r2);

	findNearestNeighbour(l1, r1, x, y);
	if (delta2 < closestDist)
		findNearestNeighbour(l2, r2, x, y);
}

int findNearestNeighbour(int n, int x, int y) {
	closestDist = LLONG_MAX;
	findNearestNeighbour(0, n, x, y);
	return closestNode;
}

int main() {
    vpii p;
    p.push_back(make_pair(0, 2));
    p.push_back(make_pair(0, 3));
    p.push_back(make_pair(-1, 0));

    p.resize(unique(p.begin(), p.end()) - p.begin());

    int n = p.size();
    buildTree(1, 0, n - 1, &(vpii(p)[0]));
    int res = findNearestNeighbour(n, 0, 0);

    cout << p[res].first << " " << p[res].second << endl;

    return 0;
}
